src/passkey_child/passkey_child_common.c

```diff
@@ -86,7 +88,9 @@ parse_arguments(int argc, const char *argv[], struct passkey_data *data)
          _("Shortname"), NULL },
         {"domain", 0, POPT_ARG_STRING, &data->domain, 0,
          _("Domain"), NULL},
-        {"public-key", 0, POPT_ARG_STRING, &data->public_key, 0,
+        {"pin", 0, POPT_ARG_STRING, &data->pin, 0,
+         _("Passkey PIN"), NULL},
```

Comment:
The calling process will prompt for the PIN when the user is authenticating and it's up to the user to provide it.

Reply:
How will the calling process know that the authenticator will require a PIN?

Reply:
Unfortunately it doesn't know it. It's the user who does. So, I think we'll need to ask for the PIN always.

Reply:
I'm keeping this point open as I need to think this through.

Reply:
I'd additionally like to make another point on this topic:

Assuming that ...
- the user has registered multiple authenticators;
- they have set different PINS on each of these authenticators; and
- have some/all plugged in during authentication.

Since you're using pre-flights, the user cannot know which authenticator will be selected by the algorithm. It is entirely dependent on which authenticator that happens to be enumerated first. They are then forced to guess which authenticator was selected first and risk burning a PIN retry.

This is why it might be preferable to have the user select an authenticator via touch whenever UV is requested by the RP.

Reply:
Another data point: passing the PIN as an argument is leaky (visible to other users monitoring processes on the system).

Reply:
> Since you're using pre-flights, the user cannot know which authenticator will be selected by the algorithm. It is entirely dependent on which authenticator that happens to be enumerated first. They are then forced to guess which authenticator was selected first and risk burning a PIN retry.

This is a limitation that we are willing to risk. pam_sss will be in charge of getting the PIN and inputting it to the helper process.

> Another data point: passing the PIN as an argument is leaky/a potential security issue (visible to other users monitoring processes on the system).

This is a more serious concern and I will bring it to our meeting.

Reply:
> > Another data point: passing the PIN as an argument is leaky/a potential security issue (visible to other users monitoring processes on the system).
> 
> This is a more serious concern and I will bring it to our meeting.

`p11_child` receives PIN via fd:
https://github.com/SSSD/sssd/blob/955192b12fafdb55daaa0b12381322031fd20816/src/p11_child/p11_child_common.c#L377

Reply:
> Another data point: passing the PIN as an argument is leaky/a potential security issue (visible to other users monitoring processes on the system).

I followed Alexey's suggestion and this is read from stdin to avoid the security issue that you reported.

Reply:
> I followed Alexey's suggestion and this is read from stdin to avoid the security issue that you reported.

:+1:

IIUC, this still means that the calling process needs to supply a PIN every time UV is requested, otherwise the process may hang indefinitely?

> This is a limitation that we are willing to risk. pam_sss will be in charge of getting the PIN and inputting it to the helper process. For that purpose pam_sss will use a PAM conversation to communicate with the user. There isn't any other way of doing it if we want to enable the authentication by different means (console, graphical interface, etc.)

I'm assuming the limitation you're referring to here is the speculative PIN prompt? There are multiple ways you could work around this; employing a simple IPC or have the passkey child return a particular error status when it needs a PIN to name a few. Both would make `pam_sss`'s PIN prompts reactive instead and users would not have to always type their PIN. Consider for example a valid fingerprint match with the Yubikey Bio. Like previously mentioned, I won't block this PR, but I highly encourage you to consider these improvements to the UX. :slightly_smiling_face:

Reply:
> IIUC, this still means that the calling process needs to supply a PIN every time UV is requested, otherwise the process may hang indefinitely?

Yes, but take into account that sssd's main process is the caller, and it'll be aware of this behaviour,

> I'm assuming the limitation you're referring to here is the speculative PIN prompt? There are multiple ways you could work around this: employing a simple IPC, signals, or having the passkey child return a particular error status when it needs a PIN, to name a few. These would make `pam_sss`'s PIN prompts reactive instead and users would not have to always type their PIN. Consider for example a valid fingerprint match with the Yubikey Bio. Like previously mentioned, I won't block this PR, but I highly encourage you to consider these improvements to the UX at some point. slightly_smiling_face

For the moment I'll keep the PR as it is, and I'll see how to improve the UX.